maitake-sync/src/wait_queue.rs

```diff
@@ -684,6 +684,56 @@ impl WaitQueue {
         }
     }
 
+    /// Asynchronously poll the [`WaitQueue`] until a condition occurs
+    ///
+    /// This can be used to implement a "wait loop", turning a "try" function
+    /// (e.g. "try_recv" or "try_send") into an asynchronous function (e.g.
+    /// "recv" or "send").
+    ///
+    /// Consider using [`Self::wait_for_value()`] if your function does return a value.
+    ///
+    /// * Returns `Ok(())` if the closure returns `true`.
+    /// * Returns `Err(Closed)` if the [`WaitQueue`] is closed.
+    pub async fn wait_for<F: FnMut() -> bool>(&self, mut f: F) -> WaitResult<()> {
+        loop {
+            let wait = self.wait();
+            let mut pwait = core::pin::pin!(wait);
+            match pwait.as_mut().subscribe() {
+                Poll::Ready(wr) => wr?,
+                Poll::Pending => {}
+            }
```

Comment:
nit, take it or leave it: I think this could be written as 
```suggestion
            let _ = pwait.as_mut().subscribe()?;
```
since `Poll<Result<T, E>>` [implements `Try<Output = Poll<T>>`](https://doc.rust-lang.org/stable/std/task/enum.Poll.html#impl-Try-for-Poll%3CResult%3CT,+E%3E%3E) and then we can just throw out the `Poll` value.
 
on the other hand, maybe the match is clearer to the reader? i could be convinced either way.